packages/core/src/SessionEventEmitter.ts

```diff
@@ -0,0 +1,309 @@
+//
+// Copyright 2022 Inrupt Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+// Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+import { EventEmitter } from "events";
+import { EVENTS } from "./constant";
+
+export interface IHasSessionEventEmitter {
+  events: ISessionEventEmitter;
+}
+
+// These types help preventing inconsistencies between on, once and off.
+type LOGIN_ARGS = { eventName: typeof EVENTS.LOGIN; listener: () => void };
+type LOGOUT_ARGS = { eventName: typeof EVENTS.LOGOUT; listener: () => void };
+type SESSION_EXPIRED_ARGS = {
+  eventName: typeof EVENTS.SESSION_EXPIRED;
+  listener: () => void;
+};
+type SESSION_RESTORED_ARGS = {
+  eventName: typeof EVENTS.SESSION_RESTORED;
+  listener: (currentUrl: string) => unknown;
+};
+type ERROR_ARGS = {
+  eventName: typeof EVENTS.ERROR;
+  listener: (error: string | null, errorDescription?: string | null) => unknown;
+};
+type SESSION_EXTENDED_ARGS = {
+  eventName: typeof EVENTS.SESSION_EXTENDED;
+  listener: (expiresIn: number) => void;
+};
+type TIMEOUT_SET_ARGS = {
+  eventName: typeof EVENTS.TIMEOUT_SET;
+  listener: (timeoutHandle: number) => void;
+};
+type NEW_REFRESH_TOKEN_ARGS = {
+  eventName: typeof EVENTS.NEW_REFRESH_TOKEN;
+  listener: (newToken: string) => void;
+};
+type FALLBACK_ARGS = {
+  eventName: Parameters<InstanceType<typeof EventEmitter>["on"]>[0];
+  // Prevents from using a SessionEventEmitter as an aritrary EventEmitter.
+  listener: never;
+};
+
+export interface ISessionEventEmitter extends EventEmitter {
+  // The overloads should not be line-separated.
+  /* eslint-disable lines-between-class-members */
+  /**
+   * Register a listener called on successful login.
+   * @param eventName The login event name
+   * @param listener The callback called on successful login
+   */
+  on(
+    eventName: LOGIN_ARGS["eventName"],
+    listener: LOGIN_ARGS["listener"]
+  ): this;
```

Comment:
It might be worthwhile overloading the `emit` event as well to do one of two things:
 - Disallow it's use entirely
 - Ensure that when used with events used by on/once/off that the arguments match the types of the arguments for the listener function.

Reply:
It might also be worth overloading the `addListener` and `removeListener` parameters which correspond to the `on`, `off` events for developers that use the old API *or* marking `addListener` and `removeListener` as deprecated methods.

Reply:
Should be all fixed in 60e368ecfe51ba971e8fe72f7c0fee84f6e28aac